# Transfer Efficiency Testing with SPORF, MORF and Honest Forest
Here we assess transfer learning efficacy of SPORF, MORF and Honest Forest model over MNIST ➔ Fashion-MNIST tasks. We concentrate on the very-small target training sizes (0.1%–1%) to identify when transfer learning works because we aiming to show that the multitask classifier can harness the source domain data to improve performance in the severe data-limiting condition of the target domain. The second objective here is to know when transfer learning significantly increases the accuracy of the predictions and in the process to give evidence that the multitask classifier is functioning.

##Package Loading

In [ ]:
!pip install git+https://github.com/neurodata/treeple.git

  Cloning https://github.com/neurodata/treeple.git to /tmp/pip-req-build-f2iz8ff4
  Running command git clone --filter=blob:none --quiet https://github.com/neurodata/treeple.git /tmp/pip-req-build-f2iz8ff4
  Resolved https://github.com/neurodata/treeple.git to commit 75c2cf919939574e4240fe261f053162039495cf
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for treeple: filename=treeple-0.10.3-cp311-cp311-linux_x86_64.whl size=2924539 sha256=3f2a0430876d0d841db7f5497ce435e5483dcfac5f87360e7ac3bc0b01e915a8
  Stored in directory: /tmp/pip-ephem-wheel-cache-qyphpzr4/wheels/87/bd/56/9c9982c5af1eb667f0421e0000d3ed4b973b71c8e8903fa6cd
Successfully built treeple


In [ ]:
!pip install proglearn

In [ ]:
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from tensorflow.keras.datasets import mnist, fashion_mnist
from treeple import ObliqueRandomForestClassifier, PatchObliqueRandomForestClassifier, HonestForestClassifier
import time
import matplotlib.pyplot as plt
import seaborn as sns

## Multitask Classfier

In [ ]:
"""
MultiTaskForestClassifier:
A unified multi-task learning wrapper for SPORF, MORF, and HonestForest.
Trains on all tasks jointly and evaluates per-task performance.
"""

class MultiTaskForestClassifier:
    def __init__(self, clf_type="SPORF", task_ratios=None, random_state=42, **kwargs):
        if clf_type == "SPORF":
            self.model_cls = ObliqueRandomForestClassifier
            self.default_params = {
                "n_estimators": 200,
                "feature_combinations": 2.0,
                "max_depth": 20,
                "min_samples_split": 5,
                "min_samples_leaf": 1,
                "max_features": 0.5,
                "bootstrap": True
            }
        elif clf_type == "MORF":
            self.model_cls = PatchObliqueRandomForestClassifier
            self.default_params = {
                "n_estimators": 200
             }
        elif clf_type == "HonestForest":
            self.model_cls = HonestForestClassifier
            self.default_params = {

                "n_estimators": 200,
                "max_depth": None,
                "min_samples_split": 2,
                "min_samples_leaf": 5,
                "bootstrap": True,
                "max_features": "sqrt"
             }
        else:
            raise ValueError(f"Unsupported tree: {clf_type}")

        self.params = {**self.default_params, **kwargs}
        self.model = None
        self.task_data = {}

        if task_ratios is None:
            warnings.warn("No task_ratios provided. Using default {0: 0.5, 1: 0.5}. Override if needed.")
            self.task_ratios = {0: 0.5, 1: 0.5}
        else:
            self.task_ratios = task_ratios


        self.random_state = random_state


    def add_task(self, task_id, X, y, test_size=0.2):
        if test_size > 0:
            X_train, X_test, y_train, y_test = train_test_split(
                X, y,
                test_size=test_size,
                stratify=y,
                random_state=self.random_state
            )
        else:
            X_train, y_train = X, y
            X_test, y_test = None, None

        self.task_data[task_id] = {
            "train": (X_train, y_train),
            "test": (X_test, y_test)
        }


    def get_task_ids(self):
      return list(self.task_data.keys())

    def fit(self):
        """Train on all tasks jointly (multi-task learning)."""
        X_all, y_all, task_labels = [], [], []
        for task_id, data in self.task_data.items():
            X, y = data["train"]
            X_all.append(X)
            y_all.append(y)
            task_labels.append(np.full(len(y), task_id))

        X_all = np.vstack(X_all)
        y_all = np.concatenate(y_all)
        task_labels = np.concatenate(task_labels)
        X_all = np.column_stack((X_all, task_labels))

        self.model = self.model_cls(**self.params, random_state=42)
        self.model.fit(X_all, y_all)

    def predict(self, X, task_id):
        """Make predictions for a specific task."""
        X_task = np.column_stack((X, np.full(len(X), task_id)))
        return self.model.predict(X_task)


    def score(self, task_id):
        """Return accuracy on the held-out test set for a specific task."""
        X_test, y_test = self.task_data[task_id]["test"]
        return accuracy_score(y_test, self.predict(X_test, task_id))

def evaluate_transfer_efficiency(
    X_source, y_source,
    X_target, y_target,
    target_ratios=[0.01],
    seed=42,
    clf_type="SPORF"  #"SPORF", "MORF", "HonestForest"
):
    """
    Evaluate transfer efficiency using the specified classifier type.
    Fixes train/test split (80/20), and varies how much target training data is used.
    """
    assert clf_type in ["SPORF", "MORF", "HonestForest"], f"Invalid clf_type: {clf_type}"
    print(f"\n=== Transfer Efficiency using {clf_type} ===")
    X_train_source, X_test_source, y_train_source, y_test_source = train_test_split(
        X_source, y_source, test_size=0.2, stratify=y_source, random_state=seed
    )
    X_train_target, X_test_target, y_train_target, y_test_target = train_test_split(
        X_target, y_target, test_size=0.2, stratify=y_target, random_state=seed
    )

    for ratio in target_ratios:
        if ratio < 1.0:
            X_train_target_sub, _, y_train_target_sub, _ = train_test_split(
                X_train_target, y_train_target,
                train_size=ratio,
                stratify=y_train_target,
                random_state=seed
            )
        else:
            X_train_target_sub, y_train_target_sub = X_train_target, y_train_target
        clf_base = MultiTaskForestClassifier(
            clf_type=clf_type,
            task_ratios={1: 1.0},
            random_state=seed
        )
        clf_base.add_task(1, X_train_target_sub, y_train_target_sub, test_size=0)
        clf_base.fit()
        acc_base = accuracy_score(y_test_target, clf_base.predict(X_test_target, task_id=1))
        clf_transfer = MultiTaskForestClassifier(
            clf_type=clf_type,
            task_ratios={0: 1.0, 1: 1.0},
            random_state=seed
        )
        clf_transfer.add_task(0, X_train_source, y_train_source, test_size=0)
        clf_transfer.add_task(1, X_train_target_sub, y_train_target_sub, test_size=0)
        clf_transfer.fit()
        acc_transfer = accuracy_score(y_test_target, clf_transfer.predict(X_test_target, task_id=1))
        num_samples = len(X_train_target_sub)
        print(f"Target Task Ratio: {ratio:.4f} ({num_samples} samples)")
        print(f"  Baseline (Source 0%):   Accuracy = {acc_base:.3f}")
        print(f"  Transfer (Source 100%): Accuracy = {acc_transfer:.3f}")
        print("-" * 50)

##SPORF using the Multitask Clf
**Task:** MNIST (classes 0 and 1) → FashionMNIST (classes 0 and 1) transfer

**Model:** SPORF (Oblique Random Forest Classifier)

**Goal:**  
- Test transfer efficiency at extremely small target data ratios: 0.1%, 0.2%, 0.5%, 1%.  
- Compare baseline (no transfer) vs. full transfer (source 100%).  
- Report both **target ratio** and **number of samples** used.

---

In [ ]:
(X_mnist_train, y_mnist_train), (X_mnist_test, y_mnist_test) = mnist.load_data()
(X_fashion_train, y_fashion_train), (X_fashion_test, y_fashion_test) = fashion_mnist.load_data()
X_mnist = np.concatenate([X_mnist_train, X_mnist_test]).reshape(-1, 28*28) / 255.0
y_mnist = np.concatenate([y_mnist_train, y_mnist_test])
X_fashion = np.concatenate([X_fashion_train, X_fashion_test]).reshape(-1, 28*28) / 255.0
y_fashion = np.concatenate([y_fashion_train, y_fashion_test])
chosen_classes = [0, 1]

mnist_mask = np.isin(y_mnist, chosen_classes)
fashion_mask = np.isin(y_fashion, chosen_classes)

X_mnist = X_mnist[mnist_mask]
y_mnist = y_mnist[mnist_mask]
X_fashion = X_fashion[fashion_mask]
y_fashion = y_fashion[fashion_mask]
y_mnist = np.where(y_mnist == chosen_classes[0], 0, 1)
y_fashion = np.where(y_fashion == chosen_classes[0], 0, 1)

subset = 10000
X_mnist, y_mnist = X_mnist[:subset], y_mnist[:subset]
X_fashion, y_fashion = X_fashion[:subset], y_fashion[:subset]

evaluate_transfer_efficiency(
    X_source=X_mnist,
    y_source=y_mnist,
    X_target=X_fashion,
    y_target=y_fashion,
    target_ratios=[0.001, 0.002, 0.005, 0.01],
    seed=42,
    clf_type="SPORF"
)


=== Transfer Efficiency using SPORF ===
Target Task Ratio: 0.0010 (8 samples)
  Baseline (Source 0%):   Accuracy = 0.918
  Transfer (Source 100%): Accuracy = 0.909
--------------------------------------------------
Target Task Ratio: 0.0020 (16 samples)
  Baseline (Source 0%):   Accuracy = 0.874
  Transfer (Source 100%): Accuracy = 0.940
--------------------------------------------------
Target Task Ratio: 0.0050 (40 samples)
  Baseline (Source 0%):   Accuracy = 0.967
  Transfer (Source 100%): Accuracy = 0.961
--------------------------------------------------
Target Task Ratio: 0.0100 (80 samples)
  Baseline (Source 0%):   Accuracy = 0.980
  Transfer (Source 100%): Accuracy = 0.971
--------------------------------------------------


###Results: Transfer Efficiency using SPORF (MNIST ➔ Fashion-MNIST)

| Target Ratio | # Samples | Baseline Accuracy (Source 0%) | Transfer Accuracy (Source 100%) | Interpretation |
|:------------:|:---------:|:-----------------------------:|:-------------------------------:|:--------------:|
| 0.1%         | 8         | 0.918                         | 0.909                           | Noise (not meaningful) |
| 0.2%         | 16        | 0.874                         | 0.940                           | Successful Transfer |
| 0.5%         | 40        | 0.967                         | 0.961                           | Noise (small fluctuation) |
| 1.0%         | 80        | 0.980                         | 0.971                           | Noise (small fluctuation) |

---

###Summary:

- At **0.2%** (16 samples), **transfer learning improves accuracy** compared to baseline.
- Other ratios (0.1%, 0.5%, 1.0%) show minor variations, but no clear benefit — mostly **noise**.
- **Conclusion**: Transfer learning using SPORF is effective when training data is extremely limited.

##MORF using the Multitask Clf
**Task:** MNIST (classes 0 and 1) → FashionMNIST (classes 0 and 1) transfer

**Model:** MORF (Patch Oblique Random Forest Classifier)

**Goal:**  
- Test transfer efficiency at extremely small target data ratios: 0.1%, 0.2%, 0.5%, 1%.  
- Compare baseline (no transfer) vs. full transfer (source 100%).  
- Report both **target ratio** and **number of samples** used.

---

In [ ]:
(X_mnist_train, y_mnist_train), (X_mnist_test, y_mnist_test) = mnist.load_data()
(X_fashion_train, y_fashion_train), (X_fashion_test, y_fashion_test) = fashion_mnist.load_data()
X_mnist = np.concatenate([X_mnist_train, X_mnist_test]).reshape(-1, 28*28) / 255.0
y_mnist = np.concatenate([y_mnist_train, y_mnist_test])
X_fashion = np.concatenate([X_fashion_train, X_fashion_test]).reshape(-1, 28*28) / 255.0
y_fashion = np.concatenate([y_fashion_train, y_fashion_test])
chosen_classes = [0, 1]

mnist_mask = np.isin(y_mnist, chosen_classes)
fashion_mask = np.isin(y_fashion, chosen_classes)

X_mnist = X_mnist[mnist_mask]
y_mnist = y_mnist[mnist_mask]
X_fashion = X_fashion[fashion_mask]
y_fashion = y_fashion[fashion_mask]
y_mnist = np.where(y_mnist == chosen_classes[0], 0, 1)
y_fashion = np.where(y_fashion == chosen_classes[0], 0, 1)

subset = 10000
X_mnist, y_mnist = X_mnist[:subset], y_mnist[:subset]
X_fashion, y_fashion = X_fashion[:subset], y_fashion[:subset]

evaluate_transfer_efficiency(
    X_source=X_mnist,
    y_source=y_mnist,
    X_target=X_fashion,
    y_target=y_fashion,
    target_ratios=[0.001, 0.002, 0.005, 0.01],
    seed=42,
    clf_type="MORF"
)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step

=== Transfer Efficiency using MORF ===
Target Task Ratio: 0.0010 (8 samples)
  Baseline (Source 0%):   Accuracy = 0.651
  Transfer (Source 100%): Accuracy = 0.648
--------------------------------------------------
Target Task Ratio: 0.0020 (16 samples)
  Baseline (Source 0%):   Accuracy = 0.648
  Transfer (Source 100%): Accuracy = 0.766
--------------------------------------------------
Target Task Ratio: 0.0050 (40 samples)
  Baseline (Source 0%):   Accuracy = 0.681
  Transfer (Source 100%): Accuracy = 0.786
--------------------------------------------------
Target Task Ratio: 0.0100 (80 samples)
  Baseline (Source 0%):   Accuracy = 0.865
  Transfer (Source 100%): Accuracy = 0.878
--------------------------------------------------


###Results: Transfer Efficiency using MORF (MNIST ➔ Fashion-MNIST)

| Target Ratio | # Samples | Baseline Accuracy (Source 0%) | Transfer Accuracy (Source 100%) | Interpretation |
|:------------:|:---------:|:-----------------------------:|:-------------------------------:|:--------------:|
| 0.1%         | 8         | 0.651                         | 0.648                           | Noise (small fluctuation) |
| 0.2%         | 16        |                 0.648         |             0.766               | Successful Transfer |
| 0.5%         | 40        |  0.681                        |   0.786                         | Successful Transfer  |
| 1.0%         | 80        |          0.865                |   0.878                         | Successful Transfer  |

---

###Summary:

- At **0.2%**, **0.5%** and **1.0%** (16, 40 and 80 samples), **transfer learning improves accuracy** compared to baseline.
- 0.1% ratio shows a minor variation, but no clear benefit — mostly **noise**.
- **Conclusion**: Transfer learning using MORF is effective when training data is extremely limited.

HONEST FOREST

In [ ]:
(X_mnist_train, y_mnist_train), (X_mnist_test, y_mnist_test) = mnist.load_data()
(X_fashion_train, y_fashion_train), (X_fashion_test, y_fashion_test) = fashion_mnist.load_data()
X_mnist = np.concatenate([X_mnist_train, X_mnist_test]).reshape(-1, 28*28) / 255.0
y_mnist = np.concatenate([y_mnist_train, y_mnist_test])
X_fashion = np.concatenate([X_fashion_train, X_fashion_test]).reshape(-1, 28*28) / 255.0
y_fashion = np.concatenate([y_fashion_train, y_fashion_test])
chosen_classes = [0, 1]

mnist_mask = np.isin(y_mnist, chosen_classes)
fashion_mask = np.isin(y_fashion, chosen_classes)

X_mnist = X_mnist[mnist_mask]
y_mnist = y_mnist[mnist_mask]
X_fashion = X_fashion[fashion_mask]
y_fashion = y_fashion[fashion_mask]
y_mnist = np.where(y_mnist == chosen_classes[0], 0, 1)
y_fashion = np.where(y_fashion == chosen_classes[0], 0, 1)

subset = 10000
X_mnist, y_mnist = X_mnist[:subset], y_mnist[:subset]
X_fashion, y_fashion = X_fashion[:subset], y_fashion[:subset]

evaluate_transfer_efficiency(
    X_source=X_mnist,
    y_source=y_mnist,
    X_target=X_fashion,
    y_target=y_fashion,
    target_ratios=[0.002, 0.005, 0.01, 0.02, 0.03, 0.04, 0.05, 0.1, 0.2, 0.5],
    seed=42,
    clf_type="HonestForest"
)


=== Transfer Efficiency using HonestForest ===
Target Task Ratio: 0.0020 (16 samples)
  Baseline (Source 0%):   Accuracy = 0.498
  Transfer (Source 100%): Accuracy = 0.782
--------------------------------------------------
Target Task Ratio: 0.0050 (40 samples)
  Baseline (Source 0%):   Accuracy = 0.947
  Transfer (Source 100%): Accuracy = 0.810
--------------------------------------------------
Target Task Ratio: 0.0100 (80 samples)
  Baseline (Source 0%):   Accuracy = 0.972
  Transfer (Source 100%): Accuracy = 0.863
--------------------------------------------------
Target Task Ratio: 0.0200 (160 samples)
  Baseline (Source 0%):   Accuracy = 0.978
  Transfer (Source 100%): Accuracy = 0.895
--------------------------------------------------
Target Task Ratio: 0.0300 (240 samples)
  Baseline (Source 0%):   Accuracy = 0.979
  Transfer (Source 100%): Accuracy = 0.910
--------------------------------------------------
Target Task Ratio: 0.0400 (320 samples)
  Baseline (Source 0%):   Accu